andy_elf/multiboot.c

```diff
@@ -0,0 +1,194 @@
+    multiboot2_late_init(magic, addr);
+    return 1; // Successfully processed multiboot2 info
+}
+
+/**
+ * Late-init for multiboot2. This function is called to get multiboot2 metadata after the core services like the memory manager are up and running.
+ */
+uint32_t multiboot2_late_init(uint32_t magic, uint32_t addr) {
+    kprintf("Checking for multiboot2 info at %x with magic %x\r\n", addr, magic);
+    // Check the magic number to ensure we were booted by a multiboot2-compliant bootloader
+    if (magic != 0x36D76289) {  //this is the magic number the bootloader passes to us
+        return 0; // Not a valid multiboot2 boot
+    }
+
+    kernel_config_ptr = new_kernel_config();
+    if(!kernel_config_ptr) {
+        k_panic("FATAL Could not allocate kernel config structure\r\n");
+        return 0;
+    }
+
+    struct MultibootKernelDataHeader *header = (struct MultibootKernelDataHeader *)addr;
+    kprintf("Multiboot2 info total size %d\r\n", header->total_size);
+
+    struct MultibootTagHeader *tag = (struct MultibootTagHeader *)(addr + sizeof(struct MultibootKernelDataHeader));
+    while (tag && tag->type != MB_TAG_END) {
+        // Process each tag as needed
+        switch (tag->type) {
+            case MB_TAG_BASIC_MEMINFO:
+                break;
+            case MB_TAG_BOOTLOADER_NAME:
+                break;
+            case MB_TAG_BOOTDEV:
+                struct MultibootTagBiosBootDevice *bootdev = (struct MultibootTagBiosBootDevice *)tag;
+                kprintf("Boot device: 0x%x:0x%x:0x%x\r\n", bootdev->biosdev, bootdev->partition, bootdev->subpartition);
+                kernel_config_ptr->bios_boot_device = bootdev->biosdev;
+                kernel_config_ptr->bios_partition = bootdev->partition;
+                kernel_config_ptr->bios_subpartition = bootdev->subpartition;
+                break;
+            case MB_TAG_CMDLINE:
+                struct MultibootTagCmdline *cmdline = (struct MultibootTagCmdline *)tag;
+                kprintf("Kernel command line: '%s'\r\n", cmdline->cmdline);
+                kernel_config_ptr->cmdline = parse_command_line(cmdline->cmdline);
+                if(kernel_config_ptr->cmdline==NULL) {
+                    kputs("WARNING Could not parse command line\r\n");
+                }
+                break;
+            case MB_TAG_MEMORY_MAP:
+                break;
+            case MB_TAG_APM:
+                struct MultibootTagAPM *apm = (struct MultibootTagAPM *)tag;
+                kprintf("APM version 0x%x, cseg 0x%x, offset 0x%x, cseg_16 0x%x, dseg 0x%x, flags 0x%x, cseg_len 0x%x, cseg_16_len 0x%x, dseg_len 0x%x\r\n",
+                        apm->version, apm->cseg, apm->offset, apm->cseg_16, apm->dseg, apm->flags, apm->cseg_len, apm->cseg_16_len, apm->dseg_len);
+                break;
+            case MB_TAG_RSDP_NEW:
+                kputs("ACPI RSDP 2.0 found\r\n");
+                load_acpi_data();
```

Comment:
TODO: Pass the RSDP pointer in directly (as an option, keep the 'seek' behaviour as the default if NULL is passed)